File: ArtFruit/Sources/AICAPIClient.swift
import Foundation

// MARK: - Models

struct AICArtwork: Identifiable {
    let id: Int
    let title: String
    let artist: String
    let imageURL: URL
}

// MARK: - Available artists (with known public-domain works in AIC and/or WikiArt)

/// Artists available in the Art Institute of Chicago collection.
/// The display name matches the `artist_title` field returned by the AIC API.
let AICAvailableArtists: [String] = [
    "Claude Monet",
    "Pierre-Auguste Renoir",
    "Paul Cézanne",
    "Edgar Degas",
    "Georges Seurat",
    "Vincent van Gogh",
    "Paul Gauguin",
    "Henri de Toulouse-Lautrec",
    "Édouard Manet",
    "Camille Pissarro",
    "Gustave Courbet",
    "Eugène Delacroix",
    "Jean-Auguste-Dominique Ingres",
    "Francisco José de Goya y Lucientes",
    "Rembrandt van Rijn",
    "Johannes Vermeer",
    "Peter Paul Rubens",
    "El Greco",
    "Caravaggio",
    "Sandro Botticelli",
    "Raphael",
    "Leonardo da Vinci",
    "Michelangelo Buonarroti",
    "Albrecht Dürer",
    "Hieronymus Bosch",
    "Jan van Eyck",
    "Pablo Picasso",
    "Henri Matisse",
    "Salvador Dalí",
    "René Magritte",
    "Wassily Kandinsky",
    "Paul Klee",
    "Piet Mondrian",
    "Amedeo Modigliani",
    "Marc Chagall",
    "Egon Schiele",
    "Gustav Klimt",
    "Edvard Munch",
    "James McNeill Whistler",
    "Winslow Homer",
    "John Singer Sargent",
    "Mary Cassatt",
    "Grant Wood",
    "Georgia O'Keeffe",
    "Edward Hopper",
    "Jackson Pollock",
    "Mark Rothko",
    "Andy Warhol",
    "Roy Lichtenstein",
    "Utagawa Hiroshige",
    "Katsushika Hokusai",
]

// MARK: - Available styles (sourced from AIC aggregations on public-domain artworks)

let AICAvailableStyles: [String] = [
    "Impressionism",
    "Post-Impressionism",
    "Surrealism",
    "Abstract Expressionism",
    "Modernism",
    "Baroque",
    "Renaissance",
    "Romanticism",
    "Realism",
    "Art Nouveau",
    "Art Deco",
    "Cubism",
    "Expressionism",
    "Minimalism",
    "Pop Art",
    "contemporary",
    "medieval",
    "19th century",
    "18th Century",
    "17th Century",
    "greek",
    "egyptian",
    "roman (ancient, style or period)",
    "Japanese (culture or style)",
    "Chinese (culture or style)",
    "South Asian",
    "Himalayan",
    "Pictorialism",
    "nazca",
    "moche",
]

// MARK: - Private response types

private struct AICListResponse: Decodable {
    struct Pagination: Decodable {
        let total: Int
        let totalPages: Int
        enum CodingKeys: String, CodingKey {
            case total
            case totalPages = "total_pages"
        }
    }
    struct ArtworkData: Decodable {
        let id: Int
        let title: String?
        let artistTitle: String?
        let imageId: String?
        enum CodingKeys: String, CodingKey {
            case id
            case title
            case artistTitle = "artist_title"
            case imageId = "image_id"
        }
    }
    let pagination: Pagination
    let data: [ArtworkData]
}

private struct AICSearchResponse: Decodable {
    struct Pagination: Decodable {
        let total: Int
        let totalPages: Int
        enum CodingKeys: String, CodingKey {
            case total
            case totalPages = "total_pages"
        }
    }
    struct ArtworkData: Decodable {
        let id: Int
        let title: String?
        let artistTitle: String?
        let imageId: String?
        enum CodingKeys: String, CodingKey {
            case id
            case title
            case artistTitle = "artist_title"
            case imageId = "image_id"
        }
    }
    let pagination: Pagination
    let data: [ArtworkData]
}

// MARK: - Client

final class AICAPIClient {
    private let baseURL = "https://api.artic.edu/api/v1"
    private let iiifBase = "https://www.artic.edu/iiif/2"
    private let session: URLSession = {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 30
        return URLSession(configuration: config)
    }()

    /// Fetch a random public-domain artwork, optionally filtered to one of the given styles
    /// and/or one of the given artists.
    /// Empty sets mean "no filter — show everything".
    func randomArtwork(styles: Set<String> = [], artists: Set<String> = []) async throws -> AICArtwork {
        if styles.isEmpty && artists.isEmpty {
            return try await randomArtworkUnfiltered()
        } else {
            return try await randomArtworkFiltered(styles: styles, artists: artists)
        }
    }

    // MARK: - Unfiltered path (original behaviour)

    private func randomArtworkUnfiltered() async throws -> AICArtwork {
        let firstPage = try await fetchArtworkPage(1)
        let totalPages = min(firstPage.pagination.totalPages, 12000)
        NSLog("[ArtFruit] Total pages: \(totalPages)")

        let randomPage = Int.random(in: 1...totalPages)
        NSLog("[ArtFruit] Fetching page \(randomPage)")

        let page = randomPage == 1 ? firstPage : (try await fetchArtworkPage(randomPage))
        let withImages = page.data.filter { $0.imageId != nil && !($0.imageId!.isEmpty) }
        NSLog("[ArtFruit] Artworks with images on this page: \(withImages.count)")

        guard let pick = withImages.randomElement(), let imageId = pick.imageId else {
            let fallback = firstPage.data.filter { $0.imageId != nil && !($0.imageId!.isEmpty) }
            guard let pick2 = fallback.randomElement(), let imageId2 = pick2.imageId else {
                throw ArtFruitError.noArtworksFound
            }
            return makeArtwork(pick2, imageId: imageId2)
        }

        return makeArtwork(pick, imageId: imageId)
    }

    private func fetchArtworkPage(_ page: Int) async throws -> AICListResponse {
        let urlStr = "\(baseURL)/artworks?page=\(page)&limit=100&fields=id,title,artist_title,image_id&is_public_domain=1"
        guard let url = URL(string: urlStr) else { throw ArtFruitError.noArtworksFound }
        NSLog("[ArtFruit] Fetching: \(urlStr)")

        let (data, response) = try await session.data(from: url)
        if let http = response as? HTTPURLResponse {
            NSLog("[ArtFruit] HTTP \(http.statusCode)")
        }

        return try JSONDecoder().decode(AICListResponse.self, from: data)
    }

    // MARK: - Style/Artist-filtered path (search endpoint)

    private func randomArtworkFiltered(styles: Set<String>, artists: Set<String>) async throws -> AICArtwork {
        // Build must-clauses for both style and artist filters
        var mustClauses: [[String: Any]] = [
            ["term": ["is_public_domain": true]],
            ["exists": ["field": "image_id"]]
        ]
        if !styles.isEmpty {
            let chosenStyle = Array(styles).randomElement()!
            mustClauses.append(["term": ["style_title.keyword": chosenStyle]])
            NSLog("[ArtFruit] AIC filtering by style: '\(chosenStyle)'")
        }
        if !artists.isEmpty {
            let chosenArtist = Array(artists).randomElement()!
            mustClauses.append(["match": ["artist_title": chosenArtist]])
            NSLog("[ArtFruit] AIC filtering by artist: '\(chosenArtist)'")
        }

        let firstPage = try await fetchSearchPage(1, mustClauses: mustClauses)
        guard firstPage.pagination.total > 0 else {
            NSLog("[ArtFruit] No results for the given filters, falling back to unfiltered.")
            return try await randomArtworkUnfiltered()
        }

        // Cap at 100 pages (10,000 results — Elasticsearch limit)
        let totalPages = min(firstPage.pagination.totalPages, 100)
        NSLog("[ArtFruit] AIC filtered: \(firstPage.pagination.total) artworks, \(totalPages) pages")

        let randomPage = Int.random(in: 1...totalPages)
        let page = randomPage == 1 ? firstPage : (try await fetchSearchPage(randomPage, mustClauses: mustClauses))

        let withImages = page.data.filter { $0.imageId != nil && !($0.imageId!.isEmpty) }
        NSLog("[ArtFruit] Artworks with images on page \(randomPage): \(withImages.count)")

        guard let pick = withImages.randomElement(), let imageId = pick.imageId else {
            let fallback = firstPage.data.filter { $0.imageId != nil && !($0.imageId!.isEmpty) }
            guard let pick2 = fallback.randomElement(), let imageId2 = pick2.imageId else {
                throw ArtFruitError.noArtworksFound
            }
            return makeArtwork(pick2, imageId: imageId2)
        }

        return makeArtwork(pick, imageId: imageId)
    }

    private func fetchSearchPage(_ page: Int, mustClauses: [[String: Any]]) async throws -> AICSearchResponse {
        let query: [String: Any] = [
            "query": [
                "bool": [
                    "must": mustClauses
                ]
            ],
            "fields": ["id", "title", "artist_title", "image_id"],
            "_source": ["id", "title", "artist_title", "image_id"],
            "from": (page - 1) * 100,
            "size": 100
        ]

        let urlStr = "\(baseURL)/artworks/search"
        guard let url = URL(string: urlStr) else { throw ArtFruitError.noArtworksFound }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.httpBody = try JSONSerialization.data(withJSONObject: query)

        NSLog("[ArtFruit] AIC search POST page \(page)")
        let (data, response) = try await session.data(for: request)
        if let http = response as? HTTPURLResponse {
            NSLog("[ArtFruit] HTTP \(http.statusCode)")
        }

        return try JSONDecoder().decode(AICSearchResponse.self, from: data)
    }

    // MARK: - Shared helpers

    private func makeArtwork(_ data: AICListResponse.ArtworkData, imageId: String) -> AICArtwork {
        let imageURL = URL(string: "\(iiifBase)/\(imageId)/full/1400,/0/default.jpg")!
        NSLog("[ArtFruit] Selected: \"\(data.title ?? "nil")\" by \(data.artistTitle ?? "nil")")
        return AICArtwork(
            id: data.id,
            title: data.title ?? "Untitled",
            artist: data.artistTitle ?? "Unknown Artist",
            imageURL: imageURL
        )
    }

    private func makeArtwork(_ data: AICSearchResponse.ArtworkData, imageId: String) -> AICArtwork {
        let imageURL = URL(string: "\(iiifBase)/\(imageId)/full/1400,/0/default.jpg")!
        NSLog("[ArtFruit] Selected: \"\(data.title ?? "nil")\" by \(data.artistTitle ?? "nil")")
        return AICArtwork(
            id: data.id,
            title: data.title ?? "Untitled",
            artist: data.artistTitle ?? "Unknown Artist",
            imageURL: imageURL
        )
    }
}

enum ArtFruitError: LocalizedError {
    case noArtworksFound

    var errorDescription: String? {
        switch self {
        case .noArtworksFound:
            return "No artworks could be found."
        }
    }
}

File: artfruit_windows/AicApiClient.cs
using System.Net.Http;
using System.Net.Http.Json;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;

namespace ArtFruit;

/// <summary>
/// Client for the Art Institute of Chicago public API. Ports the Swift
/// <c>AICAPIClient</c>: an unfiltered paginated path plus a style-filtered path
/// backed by the Elasticsearch <c>/artworks/search</c> endpoint.
/// </summary>
public sealed class AicApiClient
{
    private const string BaseUrl = "https://api.artic.edu/api/v1";
    private const string IiifBase = "https://www.artic.edu/iiif/2";

    private readonly HttpClient _http;
    private readonly Random _random = Random.Shared;

    public AicApiClient(HttpClient http)
    {
        _http = http;
    }

    /// <summary>
    /// Fetch a random public-domain artwork, optionally filtered by styles and/or
    /// artists. When both sets are empty, falls back to the unfiltered paginated listing.
    /// </summary>
    public async Task<Artwork> RandomArtworkAsync(ISet<string> styles, ISet<string> artists, CancellationToken ct = default)
    {
        if (styles.Count == 0 && artists.Count == 0)
            return await RandomArtworkUnfilteredAsync(ct).ConfigureAwait(false);

        return await RandomArtworkFilteredAsync(styles, artists, ct).ConfigureAwait(false);
    }

    // ------------------------------------------------------------------
    // Unfiltered path
    // ------------------------------------------------------------------

    private async Task<Artwork> RandomArtworkUnfilteredAsync(CancellationToken ct)
    {
        var firstPage = await FetchArtworkPageAsync(1, ct).ConfigureAwait(false);
        var totalPages = Math.Min(firstPage.Pagination.TotalPages, 12000);
        Log.Info($"Total pages: {totalPages}");

        var randomPage = _random.Next(1, Math.Max(totalPages, 1) + 1);
        Log.Info($"Fetching page {randomPage}");

        var page = randomPage == 1 ? firstPage : await FetchArtworkPageAsync(randomPage, ct).ConfigureAwait(false);
        var withImages = page.Data.Where(d => !string.IsNullOrEmpty(d.ImageId)).ToList();
        Log.Info($"Artworks with images on this page: {withImages.Count}");

        var pick = withImages.Count > 0 ? withImages[_random.Next(withImages.Count)] : null;
        if (pick is null)
        {
            var fallback = firstPage.Data.Where(d => !string.IsNullOrEmpty(d.ImageId)).ToList();
            if (fallback.Count == 0)
                throw new NoArtworksFoundException();
            var pick2 = fallback[_random.Next(fallback.Count)];
            return MakeArtwork(pick2);
        }

        return MakeArtwork(pick);
    }

    private async Task<AicListResponse> FetchArtworkPageAsync(int page, CancellationToken ct)
    {
        var url = $"{BaseUrl}/artworks?page={page}&limit=100&fields=id,title,artist_title,image_id&is_public_domain=1";
        Log.Info($"Fetching: {url}");

        using var response = await _http.GetAsync(url, ct).ConfigureAwait(false);
        Log.Info($"HTTP {(int)response.StatusCode}");
        response.EnsureSuccessStatusCode();

        var result = await response.Content.ReadFromJsonAsync<AicListResponse>(JsonOpts, ct).ConfigureAwait(false);
        return result ?? throw new NoArtworksFoundException();
    }

    // ------------------------------------------------------------------
    // Style/Artist-filtered path (Elasticsearch search endpoint)
    // ------------------------------------------------------------------

    private async Task<Artwork> RandomArtworkFilteredAsync(ISet<string> styles, ISet<string> artists, CancellationToken ct)
    {
        // Build must-clauses combining style and artist filters.
        var mustClauses = new List<object>
        {
            new { term = new Dictionary<string, object> { ["is_public_domain"] = true } },
            new { exists = new { field = "image_id" } },
        };

        if (styles.Count > 0)
        {
            var styleList = styles.ToList();
            var chosenStyle = styleList[_random.Next(styleList.Count)];
            mustClauses.Add(new { term = new Dictionary<string, object> { ["style_title.keyword"] = chosenStyle } });
            Log.Info($"AIC filtering by style: '{chosenStyle}'");
        }

        if (artists.Count > 0)
        {
            var artistList = artists.ToList();
            var chosenArtist = artistList[_random.Next(artistList.Count)];
            mustClauses.Add(new { match = new Dictionary<string, object> { ["artist_title"] = chosenArtist } });
            Log.Info($"AIC filtering by artist: '{chosenArtist}'");
        }

        var firstPage = await FetchSearchPageAsync(1, mustClauses, ct).ConfigureAwait(false);
        if (firstPage.Pagination.Total <= 0)
        {
            Log.Info("No results for the given filters, falling back to unfiltered.");
            return await RandomArtworkUnfilteredAsync(ct).ConfigureAwait(false);
        }

        // Cap at 100 pages (10,000 results — Elasticsearch limit).
        var totalPages = Math.Min(firstPage.Pagination.TotalPages, 100);
        Log.Info($"AIC filtered: {firstPage.Pagination.Total} artworks, {totalPages} pages");

        var randomPage = _random.Next(1, Math.Max(totalPages, 1) + 1);
        var page = randomPage == 1 ? firstPage : await FetchSearchPageAsync(randomPage, mustClauses, ct).ConfigureAwait(false);

        var withImages = page.Data.Where(d => !string.IsNullOrEmpty(d.ImageId)).ToList();
        Log.Info($"Artworks with images on page {randomPage}: {withImages.Count}");

        var pick = withImages.Count > 0 ? withImages[_random.Next(withImages.Count)] : null;
        if (pick is null)
        {
            var fallback = firstPage.Data.Where(d => !string.IsNullOrEmpty(d.ImageId)).ToList();
            if (fallback.Count == 0)
                throw new NoArtworksFoundException();
            var pick2 = fallback[_random.Next(fallback.Count)];
            return MakeArtwork(pick2);
        }

        return MakeArtwork(pick);
    }

    private async Task<AicListResponse> FetchSearchPageAsync(int page, List<object> mustClauses, CancellationToken ct)
    {
        var query = new
        {
            query = new
            {
                @bool = new { must = mustClauses },
            },
            fields = new[] { "id", "title", "artist_title", "image_id" },
            _source = new[] { "id", "title", "artist_title", "image_id" },
            from = (page - 1) * 100,
            size = 100,
        };

        var url = $"{BaseUrl}/artworks/search";
        var json = JsonSerializer.Serialize(query);
        using var content = new StringContent(json, Encoding.UTF8, "application/json");

        Log.Info($"AIC search POST page {page}");
        using var response = await _http.PostAsync(url, content, ct).ConfigureAwait(false);
        Log.Info($"HTTP {(int)response.StatusCode}");
        response.EnsureSuccessStatusCode();

        var result = await response.Content.ReadFromJsonAsync<AicListResponse>(JsonOpts, ct).ConfigureAwait(false);
        return result ?? throw new NoArtworksFoundException();
    }

    // ------------------------------------------------------------------
    // Shared helpers
    // ------------------------------------------------------------------

    private static Artwork MakeArtwork(AicArtworkData data)
    {
        var imageUrl = new Uri($"{IiifBase}/{data.ImageId}/full/1400,/0/default.jpg");
        Log.Info($"Selected: \"{data.Title ?? "nil"}\" by {data.ArtistTitle ?? "nil"}");
        return new Artwork(
            data.Id,
            data.Title ?? "Untitled",
            data.ArtistTitle ?? "Unknown Artist",
            imageUrl);
    }

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNameCaseInsensitive = true,
    };

    // ------------------------------------------------------------------
    // Response DTOs (both the list and search endpoints share this shape)
    // ------------------------------------------------------------------

    private sealed class AicListResponse
    {
        [JsonPropertyName("pagination")]
        public AicPagination Pagination { get; set; } = new();

        [JsonPropertyName("data")]
        public List<AicArtworkData> Data { get; set; } = new();
    }

    private sealed class AicPagination
    {
        [JsonPropertyName("total")]
        public int Total { get; set; }

        [JsonPropertyName("total_pages")]
        public int TotalPages { get; set; }
    }

    private sealed class AicArtworkData
    {
        [JsonPropertyName("id")]
        public int Id { get; set; }

        [JsonPropertyName("title")]
        public string? Title { get; set; }

        [JsonPropertyName("artist_title")]
        public string? ArtistTitle { get; set; }

        [JsonPropertyName("image_id")]
        public string? ImageId { get; set; }
    }
}
